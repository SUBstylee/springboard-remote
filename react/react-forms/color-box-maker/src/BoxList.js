import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

function BoxList(props) {
    const [boxes, setBoxes] = useState([]);

    const create = (boxObj) => {
        setBoxes(boxes => [...boxes, boxObj]);
    };

    return (
        <div>
            <h1>Color Box Maker</h1>
            <NewBoxForm createBox={create} />
            {boxes.map(box => <Box width={box.width} height={box.height} color={box.color} />)}
        </div>
    );
};

export default BoxList;