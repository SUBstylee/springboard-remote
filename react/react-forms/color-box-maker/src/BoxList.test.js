import { render, fireEvent, wait } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height = '100', width = '100', color = 'red') {
    const heightInput = boxList.getByLabelText("Height");
    const widthInput = boxList.getByLabelText('Width');
    const colorInput = boxList.getByLabelText('Color');
    fireEvent.change(colorInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("MAKE A BOX!");
    fireEvent.click(button);
}

it('renders without crashing', () => {
    render(<BoxList />);
});

it('matches snapshot', () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', () => {
    const boxList = render(<BoxList />);
    //no box on intial page load
    expect(boxList.queryByText('X')).not.toBeInTheDocument();
    //add a box
    addBox(boxList);
    //check if there is a box
    const removeBtn = boxList.getByText('X');
    expect(removeBtn).toBeInTheDocument();
    expect(removeBtn.previousSibling).toHaveStyle(`
    width: 100;
    height: 100;
    backgroundColor: red;
  `);
    //input fields should be empty after new box is made
    expect(boxList.getAllByDisplayValue('')).toHaveLength(3);
});

it('can remove a box', () => {
    const boxList = render(<BoxList />);
    addBox(boxList);
    const removeBtn = boxList.getByText('X');
    //button will remove box
    fireEvent.click(removeBtn);
    expect(removeBtn).not.toBeInTheDocument();
});
