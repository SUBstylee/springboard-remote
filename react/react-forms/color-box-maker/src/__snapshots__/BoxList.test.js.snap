// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Color Box Maker
    </h1>
    <form>
      <div>
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="color"
        >
          Color
        </label>
        <input
          id="color"
          name="color"
          type="text"
          value=""
        />
      </div>
      <button>
        MAKE A BOX!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
